File: src/clu/sources/virt_what.py
import logging

from clu.provides import Provides
from clu.requires import Requires
from clu.facts import Facts
from clu.sources import Source
from clu.input import text_program
from clu.sources import PARSE_FAIL_MSG

log = logging.getLogger(__name__)


class VirtWhat(Source):
    def provides(self, provides: Provides) -> None:
        provides["phy.platform"] = self

    def requires(self, requires: Requires) -> None:
        requires.programs.append("virt-what")

    def parse(self, facts: Facts) -> None:
        if "phy.platform" in facts:
            return

        data, rc = text_program("virt-what")
        log.debug(f"{rc=}")
        if rc != 0:
            facts["phy.platform"] = PARSE_FAIL_MSG
            return
        data = data.strip()
        data = data.replace("\n", ", ")
        log.debug(f"{data=}")
        if not data:
            data = "physical"
        facts["phy.platform"] = data

File: pkg/facts/sources/virt_what.go
package sources

// Go port of src/clu/sources/virt_what.py
// Uses `virt-what` to determine virtualization platform(s); empty output implies physical.

import (
	"strings"

	"github.com/huntermatthews/clu/pkg/facts/types"
	"github.com/huntermatthews/clu/pkg/input"
)

// VirtWhat determines phy.platform if not already set.
type VirtWhat struct{}

var platformFact = types.Fact{
	Name: "phy.platform",
	Tier: types.TierOne,
}

// Provides registers phy.platform key.
func (v *VirtWhat) Provides(p types.Provides) { p[platformFact.Name] = v }

// Requires declares program dependency.
func (v *VirtWhat) Requires(r *types.Requires) { r.Programs = append(r.Programs, "virt-what") }

// Parse sets phy.platform unless already present. Non-zero rc -> types.ParseFailMsg.
// Multiple lines are joined by ", "; empty result -> "physical".
func (v *VirtWhat) Parse(f *types.FactDB) {
	if f.Contains("phy.platform") {
		return
	}

	data, rc, _ := input.CommandRunner("virt-what")
	if rc != 0 {
		platformFact.Value = types.ParseFailMsg
		f.AddFact(platformFact)
		return
	}

	data = strings.TrimSpace(data)
	if data == "" {
		platformFact.Value = "physical"
	} else {
		// Replace newlines with comma-space
		platformFact.Value = strings.ReplaceAll(data, "\n", ", ")
	}
	f.AddFact(platformFact)
}
